### PR TITLE
drivers: can: various fixes for can_recover() and added test case

### LINKED ISSUE #44344: drivers: can: mcp2515 introduces a hard dependency on CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
**Describe the bug**
Kconfig option `CONFIG_CAN_MCP2515=y` unconditionally selects `CONFIG_CAN_AUTO_BUS_OFF_RECOVERY=y` since the MCP2515 does not support manual bus recovery out-of-the-box.

Furthermore, the signature of the `mcp2515_recover()` function is wrong. It needs to return an `int`.

**Expected behavior**
Kconfig option `CONFIG_CAN_MCP2515=y` should not select `CONFIG_CAN_AUTO_BUS_OFF_RECOVERY=y`. Instead it should return `-ENOTSUP` from `mcp2515_recover()`.

**Impact**
This issue prevents a user from setting `CONFIG_CAN_AUTO_BUS_OFF_RECOVERY=n` when using the MCP2515 along with other CAN drivers.

**Environment (please complete the following information):**
 - OS: Ubuntu Linux 20.04 LTS
 - Toolchain: Zephyr SDK v0.14.0
 - Commit SHA: 50cc558924d1a6cdaed2a0fc9efd9979aafa0417

### LINKED ISSUE #44345: drivers: can: M_CAN bus recovery function has the wrong signature
**Describe the bug**
The `int can_mcan_recover(struct can_mcan_reg *can, k_timeout_t timeout)` function is used directly as the API `.recover` function in all in-tree Bosch M_CAN driver frontends. However, the API `.recover` function should have the following signature: `typedef int (*can_recover_t)(const struct device *dev, k_timeout_t timeout)`.

**Expected behavior**
The Bosch M_CAN driver frontends needs to define their own recover function and pass in the needed arguments to the common M_CAN recover function.

**Impact**
Calling `can_recover()` on a Bosch M_CAN driver frontend causes a runtime error.

**Environment (please complete the following information):**
 - OS: Ubuntu Linux 20.04 LTS
 - Toolchain Zephyr SDK v0.14.0
 - Commit SHA: 50cc558924d1a6cdaed2a0fc9efd9979aafa0417

### LINKED ISSUE #44348: drivers: can: z_vrfy_can_recover() does not compile
**Describe the bug**
The following code has clearly never been tested. Notice the call to `z_impl_can_recover()`.

```
static inline int z_vrfy_can_recover(const struct device *dev,
                                     k_timeout_t timeout)
{
        Z_OOPS(Z_SYSCALL_OBJ(dev, K_OBJ_DRIVER_CAN));

        return z_impl_can_recover(dev, k_timeout_t timeout);
}
#include <syscalls/can_recover_mrsh.c>
```

**To Reproduce**
1. Set `CONFIG_USERSPACE=y`, `CONFIG_CAN=y`, and `CONFIG_CAN_AUTO_BUS_OFF_RECOVERY=n`.
2. Compile any of the CAN samples or tests
3. See error

**Expected behavior**
The code should compile.

**Impact**
Unable to use `CONFIG_USERSPACE=y`, `CONFIG_CAN=y`, and `CONFIG_CAN_AUTO_BUS_OFF_RECOVERY=n` combination.

**Environment (please complete the following information):**
 - OS: Ubuntu Linux 20.04 LTS
 - Toolchain: Zephyr SDK v0.14.0
 - Commit SHA: 50cc558924d1a6cdaed2a0fc9efd9979aafa0417

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
- Do not force `CONFIG_CAN_AUTO_BUS_OFF_RECOVERY=y` when MCP2515 is enabled. Instead fix the return type of the `mcp2515_recover()` function and return `-ENOTSUP. Fixes: #44344.
- Fix the signature of the CAN bus recovery functions in the Bosch M_CAN driver frontends. Fixes: #44345.
- Fix compilation (and improve checks) in z_vrfy_can_recover(). Fixes: #44348.
- Add test case for can_recover() to prevent issues like the above from resurfacing.

### CODE REVIEW

File: drivers/can/can_mcp2515.c [hunk 1]
```diff
@@ -736,10 +736,12 @@ static void mcp2515_handle_errors(const struct device *dev)
...
 
 #ifndef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
-static void mcp2515_recover(const struct device *dev, k_timeout_t timeout)
+static int mcp2515_recover(const struct device *dev, k_timeout_t timeout)
 {
 	ARG_UNUSED(dev);
 	ARG_UNUSED(timeout);
+
+	return -ENOTSUP;
 }
```
Comment (gmarull) [inline drivers/can/can_mcp2515.c:745]: note: shouldn't this function implementation be optional? if NULL, API could then return `-ENOSYS`

Comment (henrikbrixandersen) [reply drivers/can/can_mcp2515.c:745]: I would rather not make it optional. It is common functionality that _should_ be supported by all drivers/devices.

The MCP2515 is a bit special (also in this regard) as it does not offer a register setting for selecting between manual and automatic bus recovery. The datasheet says that this _can_ be implemented by hooking into the error ISR, but this is not implemented in the Zephyr MCP2515 driver (yet?).

**Review by alexanderwachter — APPROVED** [review]

**Review by karstenkoenig — APPROVED** [review]

**Comment by karstenkoenig** [comment]
For context I tried looking at how CONFIG_CAN_AUTO_BUS_OFF_RECOVERY is handled on the other controllers, and it seems ignored for mcan, after some googling around it seems automatic bus recovery is intentionally not supported by the bosch mcan controller. So while this PR correctly untangles the config option for the mcp2515 it looks like there still is a lingering situation that the mcan controllers shouldn't allow CONFIG_CAN_AUTO_BUS_OFF_RECOVERY=y. Or am I misunderstanding the config option?

**Comment by henrikbrixandersen** [comment]
> For context I tried looking at how CONFIG_CAN_AUTO_BUS_OFF_RECOVERY is handled on the other controllers, and it seems ignored for mcan, after some googling around it seems automatic bus recovery is intentionally not supported by the bosch mcan controller. So while this PR correctly untangles the config option for the mcp2515 it looks like there still is a lingering situation that the mcan controllers shouldn't allow CONFIG_CAN_AUTO_BUS_OFF_RECOVERY=y. Or am I misunderstanding the config option?

No, you are correct. This too will be addressed, but first I would like to establish test cases for the code we already have.

The way the MCP2515 driver handled this prevented a common test case, prompting me to fix that prior to introducing the test.

My plan is to move much of the CAN configuration to run-time flags/capabilities (with compile-time Kconfigs to reduce code size where it makes sense). Auto-recovery is one candidate for this, one-shot and triple-sample others. A generic, software timer-based auto-recovery functionality is also on my list.

**Comment by karstenkoenig** [comment]
> My plan is to move much of the CAN configuration to run-time flags/capabilities (with compile-time Kconfigs to reduce code size where it makes sense). Auto-recovery is one candidate for this, one-shot and triple-sample others. A generic, software timer-based auto-recovery functionality is also on my list.

Oh nice, yes a transition to capabilities sounds like it'd sort that properly.

**Review by dleach02 — APPROVED** [review]

**Pull request merged**